### DISCUSSION TITLE
How to receive messages from multiple Lora devices? Currently I can only receive messages from one Lora device

### DISCUSSION THREAD

**Discussion by chcmq** [jgromes/RadioLib#768] [Q&A]
How to receive messages from multiple Lora devices? Currently I can only receive messages from one Lora device

```c++
void loop() {
  // check if the previous operation finished
  if(operationDone) {
    // reset flag
    operationDone = false;

    if(transmitFlag) {
      // the previous operation was transmission, listen for response
      // print the result
      if (transmissionState == RADIOLIB_ERR_NONE) {
        // packet was successfully sent
        Serial.println(F("transmission finished!"));

      } else {
        Serial.print(F("failed, code "));
        Serial.println(transmissionState);

      }

      // listen for response
      radio.startReceive();
      transmitFlag = false;

    } else {
      // the previous operation was reception
      // print data and send another packet
      String str;
      int state = radio.readData(str);

      if (state == RADIOLIB_ERR_NONE) {
        // packet was successfully received
        Serial.println(F("[SX1262] Received packet!"));

        // print data of the packet
        Serial.print(F("[SX1262] Data:\t\t"));
        Serial.println(str);

        // print RSSI (Received Signal Strength Indicator)
        Serial.print(F("[SX1262] RSSI:\t\t"));
        Serial.print(radio.getRSSI());
        Serial.println(F(" dBm"));

        // print SNR (Signal-to-Noise Ratio)
        Serial.print(F("[SX1262] SNR:\t\t"));
        Serial.print(radio.getSNR());
        Serial.println(F(" dB"));

      }

      // wait a second before transmitting again
      delay(1000);

      // send another one
      Serial.print(F("[SX1262] Sending another packet ... "));
      transmissionState = radio.startTransmit("Hello World!");
      transmitFlag = true;
    }
  
  }
}
```

**Comment by jgromes** [comment]
> Currently I can only receive messages from one Lora device

What makes you think that? You can receive from any number of transmitters you want, provided they don't transmit at the same time.

Not an issue, converting to discussion.

**Reply by chcmq** [reply]
I need to collect data of multiple motors, each motor is equipped with a Lora device, and the data of the motor is reported every three seconds, but I only have one Lora device responsible for receiving

**Reply by jgromes** [reply]
Then you need to make sure the transmitters don't transmit at the same time. For example, each transmitter can check if someone else is currently transmitting (by using channel activity detection), and if not, it can start transmitting.

**Reply by chcmq** [reply]
Can you help with a case? I need it very much, thanks

**Reply by chcmq** [reply]
void ICACHE_FLASH_ATTR lora_service::onSend(String str)
{

   // start scanning current channel
  int state = radio.scanChannel();

  if (state == RADIOLIB_LORA_DETECTED) {
    // LoRa preamble was detected
    Serial.println(F("detected!"));

  } else if (state == RADIOLIB_CHANNEL_FREE) {
    // no preamble was detected, channel is free
    Serial.println(F("channel is free!"));
    // send another one
    Serial.print(F("[SX1262] Sending another packet is: "));
    Serial.println(str);
    transmissionState = radio.startTransmit(str);
    transmitFlag = true;

  } else {
    // some other error occurred
    Serial.print(F("failed, code "));
    Serial.println(state);

  }

}
This way of writing can indeed monitor whether a message is being sent, but the receiver cannot receive the message

**Reply by jgromes** [reply]
> the receiver cannot receive the message

From the snippet above, I can already tell you are mixing together blocking `scanChannel()` with interrupt-driven `startTransmit()` and calling both from inside an interrupt service routine. You haven't said what the transmitter is doing, but I don't think it's working correctly. Once you get inside the ISR, you call scanChannel, which will raise the interrupt pin upon completion, leading to another interrupt, this time being triggered from within an interrupt context. What happens next depends on your platform. Maybe it supports nested interrupts, maybe not.

But the point is the following: the library gives you the tools, it's up to you to use them and build the application.
